### DISCUSSION TITLE
I want to run Firefox fully sandboxed ....No access to even ~/Downloads but with addons

### DISCUSSION THREAD

**Discussion by firejailaddssecuirty** [netblue30/firejail#4720] [Q&A] [answered]
I want to run Firefox fully sandboxed. No access to even ~/Downloads but with my installed addons like uBlock Origin & Noscript.
**I don't want this to be the default configuration. This is for browsing specific websites only.** 
When I use --private Firefox starts without any addons. How do I do this ?

**Comment by rusty-snake** [comment, marked as answer]
> I want to run Firefox fully sandboxed.

https://www.tecmint.com/install-and-configure-kvm-in-linux/

> I don't want this to be the default configuration. This is for browsing specific websites only.

Rethink your thread model. If you use the same firefox profile for the "normal" and the "secure" mode, where's the hole?

> No access to even ~/Downloads but with my installed addons like uBlock Origin & Noscript.

`firejail --nowhitelist='${DOWNLOADS}' /usr/bin/firefox`

**Reply by firejailaddssecuirty** [reply]
> Rethink your thread model. If you use the same firefox profile for the "normal" and the "secure" mode, where's the hole?

No, I am not using the same profile. I have 2 profiles. I have edited the Firefox desktop launcher & added "-P". Now when I click on the Firefox icon the "choose your profile" window appears. 

Okay so I will have to create a second launcher for Firefox & edit the exec line like this ?

`firejail --nowhitelist='${DOWNLOADS}' /usr/bin/firefox -P`

**Reply by rusty-snake** [reply]
> No, I am not using the same profile. I have 2 profiles. I have edited the Firefox desktop launcher & added "-P". Now when I click on the Firefox icon the "choose your profile" window appears.

Does not really change anything, both profiles are still accessible.

> Okay so I will have to create a second launcher for Firefox & edit the exec line like this ?
> 
> firejail --nowhitelist='${DOWNLOADS}' /usr/bin/firefox -P

For example (I'm not sure about quoting inside desktop files).

**Reply by firejailaddssecuirty** [reply]
> Does not really change anything, both profiles are still accessible.

Wait a second ! Suppose I create 2 Firefox profiles & name them PROFILE1 & PROFILE2. Then I choose PROFILE1 using the profile selector window. Are you saying that PROFILE1 has access to PROFILE2? I mean when PROFILE1 is in use and suppose it gets infected with something malicious PROFILE2 will get affected too ?

Then what's the point in having multiple profiles ? Why has Mozilla introduced this feature ?

**Reply by rusty-snake** [reply]
> Then what's the point in having multiple profiles ? Why has Mozilla introduced this feature ?

- Separated bookmarks/history/cookies/extensions/... so you can have a Personal profile, a Work profile and a Unbreak profile (for example).
- You can run different firefox versions at the same time
- ...

> Are you saying that PROFILE1 has access to PROFILE2?

No a profile is a passive thing (it is only a directory). But the running process has. How else did you thought a "profile selector window" is implemented?

**Reply by firejailaddssecuirty** [reply]
> No a profile is a passive thing (it is only a directory). **But the running process has. How else did you thought a "profile selector window" is implemented?**


I understand. Suppose I want to browse some websites which I do not trust at all. What should I do ? How can firejail help me in this situation ? I don't browse these websites daily. As I said if I use the --private option FF opens without any addons and without uBlock Origin I get all kind of annoying ads. Please give me an idea.

Since I don't trust these websites I don't want even ~/Downloads to be accessible.

**Reply by rusty-snake** [reply]
> Suppose I want to browse some websites which I do not trust at all. 

What kind of not trust at all? Privacy? Security? If you expect this site to exploit a vulnerability in your browser, do not visit it a block it (e.g. with ublock origin or /etc/hosts).

> As I said if I use the --private option FF opens without any addons and without uBlock Origin I get all kind of annoying ads. Please give me an idea.

Use a persistent private home: `firejail --private=${HOME}/foobar/firefox /usr/bin/firefox` (you need to configure (e.g. install uB) it once).

> Since I don't trust these websites I don't want even ~/Downloads to be accessible.

What's so private in ~/Downloads? Note that there is no full isolation if with `--private` from above you can escape the sandbox.

**Reply by firejailaddssecuirty** [reply]
> What kind of not trust at all? Privacy? Security? If you expect this site to exploit a vulnerability in your browser, do not visit it a block it (e.g. with ublock origin or /etc/hosts).

Both Privacy & Security. These are free movie streaming sites. I need to visit these websites & that's the problem. 

> Use a persistent private home: `firejail --private=${HOME}/foobar/firefox /usr/bin/firefox` (you need to configure (e.g. install uB) it once).

Okay I am going to try this right now. 

> What's so private in ~/Downloads? Note that there is no full isolation if with `--private` from above you can escape the sandbox.

Well the thing is when I download files all of them get stored in ~/Downloads & I am too lazy to move them immediately so they stay there for some time. Also if I want to upload a screenshot in my distro forum for example I have to copy the png file to ~/Downloads. Despite the fact that there is nothing that is too private in ~/Downloads I don't want to lower my security/privacy level.

**Reply by firejailaddssecuirty** [reply]
```
$ firejail --private=${HOME}/foobar/firefox /usr/bin/firefox
Error: invalid private directory
```

I created a hidden directory and named it .two & tried 

```
$ firejail --private=${HOME}/home/home/.two firefox /usr/bin/firefox
Error: invalid private directory
```
What's going wrong ?

**Reply by rusty-snake** [reply]
You need to specify a existing directory.

**Reply by firejailaddssecuirty** [reply]
```
$ pwd
/home/home/.two
```
```
$ firejail --private=${HOME}/home/home/.two /usr/bin/firefox
Error: invalid private directory
```

**Reply by rusty-snake** [reply]
`~/home/home/.two` is not a existing directory (or maybe it exists but has wrong permissions).

If you want to use `/home/home/.two`, then you need to specify it (`--private=/home/home/.two`).

**Reply by firejailaddssecuirty** [reply]
`firejail --private=/home/home/.two /usr/bin/firefox`

^^ This worked but there is one issue. When my regular Firefox is running inside firejail if I use that command its launching the same profile which is already running. I see the same bookmarks & everything else. When I close my regular Firefox window & then use that command I see a brand new profile to which I added uBlock origin. I guess this is limitation of firejail & nothing can be done. 

> Note that there is no full isolation if with --private from above you can escape the sandbox.

When you have time please explain what you mean by "you can escape the sandbox".

**Reply by rusty-snake** [reply]
> When my regular Firefox is running inside firejail if I use that command its launching the same profile which is already running. I see the same bookmarks & everything else. When I close my regular Firefox window & then use that command I see a brand new profile to which I added uBlock origin. I guess this is limitation of firejail & nothing can be done.

`firejail --private=/home/home/.two /usr/bin/firefox --no-remote`

> When you have time please explain what you mean by "you can escape the sandbox".

If you can execute arbitrary code inside the sandbox, you can (in theory) if a way out of it.
Since I (and others) already wrote about this multiple times, here are a few links:

- #3082
- https://github.com/netblue30/firejail/pull/3849#issuecomment-752230829
- https://github.com/netblue30/firejail/issues/4210#issuecomment-841882340
- https://github.com/netblue30/firejail/issues/3527#issuecomment-660686398
- #3046
